Edit contribution page (#3369)
qgis/QGIS-Documentation: source/docs/documentation_guidelines/first_contribution.rst

.. note:: Though QGIS-Documentation is used to demonstrate the process, all
   commands and steps shown below also apply to QGIS-Website.

Now that you know how to write documentation using :ref:`reStructuredText and
Sphinx<writing_doc_guidelines>`, let's dive into the process of producing

> documentation and how to quickly and safely share your changes with the

Reviewer: ```suggestion
documentation and how to share your changes quickly and safely with the
```

community.

Using the GitHub web interface
==============================

> The GitHub web interface offers you ways to easily:

Reviewer: ```suggestion
The GitHub web interface offers you easy ways to:
```

> * edit files,

Reviewer: ```suggestion
* edit files
```

> * preview and commit your changes,

Reviewer: ```suggestion
* preview and commit your changes
```

> * make a pull request to have your changes inserted into the main repository,

Reviewer: ```suggestion
* make a pull request to have your changes inserted into the main repository
```

> * and create, update, or delete branches.

Reviewer: ```suggestion
* create, update, or delete branches
```

Read the GitHub `Hello-world <https://guides.github.com/activities/hello-world/>`_
project to learn some basic vocabulary and actions that will be used below.

.. note:: **If you are fixing an issue**

    If you are making changes to fix an `issue <https://github.com/qgis/QGIS-

>     Documentation/issues>`_, add a comment to the issue to assign it to

Reviewer 2: ```suggestion
    Documentation/issues>`_, add a comment to the issue report to assign it to
```

without any harm.

Alternative 1: Using the ``Fix Me`` shortcut
............................................

Pages on the QGIS website can be edited quickly and easily by clicking on the
``Fix Me`` link at the bottom of each page.

#. This will open the file in the ``qgis:master`` branch with a message at the
   top of the page telling you that you don't have write access to this repo
   and your changes will be applied to a new branch of your repository.

#. Do your changes following the :ref:`writing guidelines <QGIS-documentation-guidelines>`.

#. When you finish, make a short comment about your changes and click on
   :guilabel:`Propose file change`. This will generate a
   new `branch <https://help.github.com/articles/about-branches/>`_ (``patch-xxx``) in your repository.

#. To contribute your changes to ``qgis/QGIS-Documentation``, skip
   to :ref:`Compare changes <compare_changes>` in the :ref:`Share your changes
   via Pull Request <sharing_changes>` section below.

Alternative 2: Edit files in your documentation repository
..........................................................

You can edit files directly from your fork of the QGIS Documentation.

> Continue with the next step.

Reviewer 2: Is this necessary?

3. Create an ad hoc branch
--------------------------

> Click on |githubBranch| in the upper left corner of your copy QGIS-

Reviewer: ```suggestion
Click on |githubBranch| in the upper left corner of your forked QGIS-
```

new `branch <https://help.github.com/articles/about-branches/>`_ .
The name of the new branch should relate to the problem you intend to fix.

.. tip:: **Do your changes in an ad hoc branch, never in the** ``master`` **branch**

   By convention, avoid making changes in your ``master`` branch except when
   you merge the modifications from the ``master`` branch of ``qgis/QGIS-Documentation``
   into your copy of the QGIS-Documentation repository.
   Separate branches allow you to work on multiple problems at the same time
   without interfering with other branches. If you make a mistake you can
   always delete a branch and start over by creating a new one from the master
   branch.

4. Make changes
---------------

#. Browse the source files to the file that you want to change.
#. Make your modifications following the :ref:`writing guidelines <QGIS-documentation-guidelines>`.
#. When you finish, navigate to the **Commit Changes** frame at the bottom of
   the page, make a short comment about your changes, and click on
   :guilabel:`Commit Changes` to commit the changes directly to your branch.
#. Repeat the previous steps for any other file that needs to be updated to

>    fix the issue.

Reviewer 2: We do agree that when taking the "fixMe" route you may not need to update more than one file before opening the PR? It obviously makes guidelines simpler but... OR can be addressed in a later stage?

Author: @DelazJ 
Yes, when you use a "Fix me" link, after you click on "Propose file change" it automatically takes you to the "Comparing changes" page. So it basically forces the user to make a separate PR for each page they edit.

Reviewer 2: Yeah, I think this is coherent with GH logic: if you create a new branch by committing, it proposes you to create a PR in the original repo (reason why users may get a compare with their own master if they use their branch as basis). But it does not necessarily need that the issue is fixed with that single file commit, in both cases (using fixme or own branch). So the step of finding and updating more files is not necessarily only for people not using the fixme option.
Anyway, this is not a big deal yet and as addressing it would/could require more restructuring, we can postpone it. let's do it step by step.

Author: Maybe the best way is to include a comment that the "fix me" link should only be used for correcting problems on one page. If they want to fix multiple pages then use the other option.

Reviewer 2: No. I'm not sure we should discourage from using the fix me link. This is the best way to ensure that a branch is up to date, and the less hassle beginners have with git and branches, the better. Their master branch could quickly be old (and/or conflictual) without a proper git app. 
The trick is "simply" not to advise to create pull request when the page shows up. 
It would be a kind of :"if all the changes required to fix the issue are done and to contribute your changes to ``qgis/QGIS-Documentation``, skip to :ref:`Compare changes <compare_changes>` in the :ref:`Share your changes via Pull Request <sharing_changes>` section below.
If the issue requires you to update other file(s), follow step YY" where YY shows how to: 
- move to their autogenerated branch (missing instruction)
- and browse files, update, commit (ie 4. Make changes)

.. note:: **If you used a** ``Fix Me`` **link**

  After you commit your changes GitHub will automatically open a new page
  comparing the changes you made in your ``patch-xxx`` branch to the ``qgis/QGIS-Documentation``
  master branch.

  Skip to :ref:`Step 2 <compare_changes>` below.

1. Start a new pull request
...........................

Navigate to the main page of the QGIS-Documentation repository (either your
fork or the main ``qgis/QGIS-Documentation`` repository) and click on
:guilabel:`New pull request`.

.. _compare_changes:

2. Compare changes
..................

If you see two dialog boxes, one that says ``base:master`` and the other
``compare:branch_name`` (see figure), this will only merge your changes from
one of your branches to your master branch. To fix this click on the
:guilabel:`compare across forks` link.

.. figure:: img/githubCompareAcrossForks.png
    :align: Center

    If your :guilabel:`Comparing changes` page looks like this, click on the
    :guilabel:`compare across forks` link.

There are four drop-down menus which allow you to compare the changes you
have made in your branch with the master branch that you want to merge into.
They are:

> * **base fork**: the fork that you want merge your changes into,

Reviewer: ```suggestion
* **base fork**: the fork that you want to merge your changes into,
```

> * **base**: the branch of the base fork that you want to merge your changes into,

Reviewer: ```suggestion
* **base**: the branch of the base fork that you want to merge your changes into
```

> * **head fork**: the fork that has changes that you want to incorporate into the base fork,

Reviewer: ```suggestion
* **head fork**: the fork that has changes that you want to incorporate into the base fork
```

> * **compare**: the branch with those changes.

Reviewer: ```suggestion
* **compare**: the branch with those changes
```

A green check with the words **Able to merge** shows that your changes can
be merged into the official documentation without conflicts.

Click the :guilabel:`Create pull request` button.

.. warning:: **If you get a red X**

    If you get a red X, it means that there are conflicts. The files
    that you are modifying are not up to date with the branch you are
    targeting because someone else has made a commit that conflicts with your
    changes. You can still create the pull request but you'll need to fix any
    :ref:`conflicts <fix_conflicts>` to complete the merge.

.. tip:: Though released and being translated, the documentation of QGIS
      |CURRENT| is still maintained and existing issues are fixed. If you are
      fixing issues for a different release, change **base** from ``master``
      to the appropriate ``release_...`` branch in the steps above.

.. warning:: **Do not make a pull request against your own** ``master`` **branch.**

    If you make a pull request against your own master branch, no one will
    know of your changes and you may pollute your own master branch. See the
    figure above.

3. Enter comments
.................

> A text box will open; fill in any relevant comments for the issue you are

Reviewer: ```suggestion
A text box will open: fill in any relevant comments for the issue you are
```

Click on the red cross or on ``Details`` in the summary section at the bottom
of the pull request page to see the details of the error. You'll have to fix
any reported errors or warnings before your changes are committed to the
``qgis/QGIS-Documentation`` repository.

    .. %ToDo: The FAQ could add information on Travis error and common fixes.

You can make modifications to your pull request until it is merged with the
main repository, either to improve your request, to address requested
modifications, or to fix a build error.

To make changes click on the |githubFilesChanged| tab in your pull request
page and click the pencil button |githubEditPencil| next to the filename that
you want to modify.

Any additional changes will be automatically added to your pull request if you
make those changes to the same branch that you submitted in your pull request.
For this reason, you should only make additional changes if those changes
relate to the issue that you intend to fix with that pull request.

If you want to fix another issue, create a new branch for
those changes and repeat the steps above.

An administrator will merge your contribution after any build errors are

> corrected, and after everything looks good to the administrators and you are

Reviewer: Do you mean "they" rather than "you"?

Author: Both. The current page says, "Once everything looks good to you and others, a committer can merge your branch with the main repo."

I was trying to emphasize that the merge would be made after the checks passed with Travis CI, after the administrators were happy and after the person submitting the changes was happy. 

Maybe a better solution is:

> An administrator will merge your contribution after any build errors are
> corrected, and after you and the administrators are
> satisfied with your changes.

satisfied with your changes.

6. Delete your ad hoc branch
----------------------------

You can delete the ad hoc branch after your changes have been merged.
Deleting old branches saves you from having unused and outdated branches in
your repository.

Navigate to your fork of the QGIS-Documentation repository (``https://github.com/<YourName>/QGIS-Documentation``).
Click on the :guilabel:`Branches` tab. Below :guilabel:`Your branches` you'll

> see a list of your branches. Click on the red trash can icon to delete any

Reviewer 2: Maybe can we reuse our |deleteSelected| icon here
```suggestion
see a list of your branches. Click on the |deleteSelected| :sup:`Delete this branch` icon to delete any
```

Author: @DelazJ 

I built it with the :sup:`Delete this branch` and the superscript text looks out of place. The red trash can icon is an improvement though.

<img width="342" alt="screen shot 2019-01-17 at 1 43 09 pm" src="https://user-images.githubusercontent.com/8138169/51341129-9b85dc80-1a5e-11e9-82df-18be6ca0ed4a.png">

Maybe this instead

Click on the |deleteSelected| icon to delete any...

Removing the term "trash can" is probably a good idea since even American english speakers don't agree on what to call that and english speakers in other countries use different terms.

Reviewer 2: What do you mean by "the superscript text looks out of place"? Your screenshot looks good to me. The :sup: role places the text by exponent and we use it when the text is a tooltip of the button, like in the GH branches page. But not problem if you prefer without; the main information is in the icon (until it changes).

unwanted branches.

.. _git_command_line_tools:

Using Git command line tools
=============================

The GitHub web interface is an easy way to update the QGIS-documentation repo with your contributions, but it doesn't offer tools to:

> * group your commits and clean your change history,

Reviewer: ```suggestion
* group your commits and clean your change history
```

> * fix possible conflicts with the main repo,

Reviewer: ```suggestion
* fix possible conflicts with the main repo
```

> * or, build the documentation to test your changes.

Reviewer: ```suggestion
* build the documentation to test your changes
```

> You then need to `install git <https://git-scm.com/downloads>`_ on your hard

Reviewer: ```suggestion
You need to `install git <https://git-scm.com/downloads>`_ on your hard
```
